### PR TITLE
Read coredata version from /.xccurrentversion file

### PR DESCRIPTION
**Pull request by matiasvillaverde**
#1540 
# Short description 📝

Currently we need to hard code the coredata model version on the manifest. This creates a duplication of setting the current version, one in Xcode, and one in Tuist. 

As a consequence, could happen that we bump the core data model on Xcode but this will not be reflected on the manifest. Therefore this PR make the currentVersion optional, and in case this is nil, we will read the currentVersion from the plist /.xccurrentversion. 

### Solution 📦

In case the current version is not explicitly set, we will try to find it on /.xccurrentversion file using `CoreDataVersionExtractor`

### Implementation 👩‍💻👨‍💻

> Detail in a checklist the steps that you took to implement the PR.

- [ ] Make currentVersion optional on ProjectDescription.CoreDataModel
- [ ] Add an extractor of the version in the mapper TuistCore.CoreDataModel

This PR was a pairing with my collague [OK](https://github.com/omkar-khedekar)

### CODE REVIEW

File: Sources/TuistLoader/Models+ManifestMappers/CoreDataModel+ManifestMapper.swift [hunk 1]
```diff
@@ -12,7 +12,63 @@ extension TuistCore.CoreDataModel {
     static func from(manifest: ProjectDescription.CoreDataModel, generatorPaths: GeneratorPaths) throws -> TuistCore.CoreDataModel {
         let modelPath = try generatorPaths.resolve(path: manifest.path)
         let versions = FileHandler.shared.glob(modelPath, glob: "*.xcdatamodel")
-        let currentVersion = manifest.currentVersion
+
+        var currentVersion: String!
```
Comment (pepibumur) [inline Sources/TuistLoader/Models+ManifestMappers/CoreDataModel+ManifestMapper.swift:16]: I didn't know that the current version is define inside the model. In that case I'd consider deprecating passing the version through the manifest. What about extending [this linter](https://github.com/tuist/tuist/blob/master/Sources/TuistGenerator/Linter/TargetLinter.swift#L25) to print a warning if the value of `currentVersion` is not nil? We can delete it in a few versions.

File: Sources/TuistLoader/Models+ManifestMappers/CoreDataModel+ManifestMapper.swift [hunk 2]
```diff
@@ -12,7 +12,63 @@ extension TuistCore.CoreDataModel {
...
+            currentVersion = hardcodedVersion
+        } else {
+            currentVersion = try CoreDataVersionExtractor.version(fromVersionFileAtPath: modelPath)
+        }
         return CoreDataModel(path: modelPath, versions: versions, currentVersion: currentVersion)
     }
 }
+
+/// Extract version from .xccurrentversion file.
+private final class CoreDataVersionExtractor {
```
Comment (pepibumur) [inline Sources/TuistLoader/Models+ManifestMappers/CoreDataModel+ManifestMapper.swift:27]: This is a good candidate to be in `TuistSupport`. What about extracting it there into its own file `CoreDataVersionExtractor`?

File: Sources/TuistLoader/Models+ManifestMappers/CoreDataModel+ManifestMapper.swift [hunk 3]
```diff
@@ -12,7 +12,63 @@ extension TuistCore.CoreDataModel {
...
+
+    /// Extract version from .xccurrentversion file
+    /// - Parameter filePath: absolute path to Model.xcdatamodel
+    /// - Throws: In case can not find the .xcurrentversion file.
+    /// - Returns: Current version of data model.
+    static func version(fromVersionFileAtPath filePath: AbsolutePath) throws -> String {
+        let url = URL(fileURLWithPath: filePath.pathString + "/.xccurrentversion")
```
Comment (pepibumur) [inline Sources/TuistLoader/Models+ManifestMappers/CoreDataModel+ManifestMapper.swift:34]: There's no need to turn `AbsolutePath` into another type that represents the same thing, `URL`.

File: Sources/TuistLoader/Models+ManifestMappers/CoreDataModel+ManifestMapper.swift [hunk 4]
```diff
@@ -12,7 +12,63 @@ extension TuistCore.CoreDataModel {
...
+        do {
+            let data = try Data(contentsOf: url)
+            let decoder = PropertyListDecoder()
+            let dict = try decoder.decode([String: String].self, from: data)
+            guard
+                let currentVersionName = dict["_XCCurrentVersionName"],
+                let versionNumberString = currentVersionName.split(separator: ".").first
```
Comment (pepibumur) [inline Sources/TuistLoader/Models+ManifestMappers/CoreDataModel+ManifestMapper.swift:41]: Why is this necessary? Does Xcode require to follow a specific convention?

File: Sources/TuistLoader/Models+ManifestMappers/CoreDataModel+ManifestMapper.swift (hunk 1 above)
Comment (kwridan) [reply Sources/TuistLoader/Models+ManifestMappers/CoreDataModel+ManifestMapper.swift:16]: For linting, seeing the model will continue to have a version, the `ManifestLinter` is probably more suited to produce this warning 🙂


https://github.com/tuist/tuist/blob/master/Sources/TuistLoader/Linter/ManifestLinter.swift

File: Sources/TuistLoader/Models+ManifestMappers/CoreDataModel+ManifestMapper.swift (hunk 4 above)
Comment (matiasvillaverde) [reply Sources/TuistLoader/Models+ManifestMappers/CoreDataModel+ManifestMapper.swift:41]: So the version is with the format `1.xccurrentversion ` this line removes the extension, so we get only the number
